programs/drift/src/instructions/admin.rs

```diff
@@ -4970,6 +4978,291 @@ pub fn handle_add_amm_constituent_data<'info>(
+    let constituent_in_token_account = &ctx.accounts.constituent_in_token_account;
+    let constituent_out_token_account = &ctx.accounts.constituent_out_token_account;
+
+    validate!(
+        amount_in <= constituent_in_token_account.amount,
+        ErrorCode::InvalidSwap,
+        "trying to swap more than the balance of the constituent in token account"
+    )?;
+
+    validate!(
+        out_constituent.flash_loan_initial_token_amount == 0,
+        ErrorCode::InvalidSwap,
+        "begin_lp_swap ended in invalid state"
+    )?;
+
+    in_constituent.flash_loan_initial_token_amount = ctx.accounts.signer_in_token_account.amount;
+    out_constituent.flash_loan_initial_token_amount = ctx.accounts.signer_out_token_account.amount;
+
+    send_from_program_vault(
+        &ctx.accounts.token_program,
+        constituent_in_token_account,
+        &ctx.accounts.signer_in_token_account,
+        &ctx.accounts.drift_signer.to_account_info(),
+        state.signer_nonce,
+        amount_in,
+        &Some(mint),
+    )?;
+
+    // The only other drift program allowed is SwapEnd
+    let mut index = current_index + 1;
+    let mut found_end = false;
+    loop {
+        let ix = match instructions::load_instruction_at_checked(index, ixs) {
+            Ok(ix) => ix,
+            Err(ProgramError::InvalidArgument) => break,
+            Err(e) => return Err(e.into()),
+        };
+
+        // Check that the drift program key is not used
+        if ix.program_id == crate::id() {
+            // must be the last ix -- this could possibly be relaxed
+            validate!(
+                !found_end,
+                ErrorCode::InvalidSwap,
+                "the transaction must not contain a Drift instruction after FlashLoanEnd"
+            )?;
+            found_end = true;
+
+            // must be the SwapEnd instruction
+            let discriminator = crate::instruction::EndLpSwap::discriminator();
+            validate!(
+                ix.data[0..8] == discriminator,
+                ErrorCode::InvalidSwap,
+                "last drift ix must be end of swap"
+            )?;
+
+            validate!(
+                ctx.accounts.signer_out_token_account.key() == ix.accounts[2].pubkey,
+                ErrorCode::InvalidSwap,
+                "the out_token_account passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.signer_in_token_account.key() == ix.accounts[3].pubkey,
+                ErrorCode::InvalidSwap,
+                "the in_token_account passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.constituent_out_token_account.key() == ix.accounts[4].pubkey,
+                ErrorCode::InvalidSwap,
+                "the in_token_account passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.constituent_in_token_account.key() == ix.accounts[5].pubkey,
```

Comment:
why dont you validate the rest of the accounts match? look the constituent and lp accounts